### PR TITLE
Add support for GPU sharing on GKE

### PR DESCRIPTION
**Pull request by grac3gao**
Implement GKE GPU sharing config settings.
fix: https://github.com/hashicorp/terraform-provider-google/issues/12679

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added `node_config.0.guest_accelerator.0.gpu_sharing_config` field to `google_container_node_pool` resource
```

### CODE REVIEW

File: mmv1/third_party/terraform/tests/resource_container_node_pool_test.go.erb [hunk 1]
```diff
@@ -2274,6 +2274,10 @@ resource "google_container_node_pool" "np_with_gpu" {
       type  = "nvidia-tesla-a100"
       gpu_partition_size = "1g.5gb"
       count = 1
+	  gpu_sharing_config {
+		gpu_sharing_strategy = "TIME_SHARING"
+		max_shared_clients_per_gpu = 2
+	  }
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/tests/resource_container_node_pool_test.go.erb:2280]: spacing appears out of place. Blocks declaration could be in inline with the rest of the variable declarations.

File: mmv1/third_party/terraform/utils/node_config.go.erb [hunk 2]
```diff
@@ -491,11 +515,23 @@ func expandNodeConfig(v interface{}) *container.NodeConfig {
...
-			guestAccelerators = append(guestAccelerators, &container.AcceleratorConfig{
+			guestAcceleratorConfig := &container.AcceleratorConfig{
 				AcceleratorCount: int64(data["count"].(int)),
 				AcceleratorType:  data["type"].(string),
 				GpuPartitionSize: data["gpu_partition_size"].(string),
-			})
+			}
+
+			if v, ok := data["gpu_sharing_config"]; ok {
+				if len(v.([]interface{})) > 0 {
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/utils/node_config.go.erb:525]: these two if statements can be condensed
```golang
if v, ok := data["gpu_sharing_config"]; ok && len(v.([]interface{})) > 0 {
```

File: mmv1/third_party/terraform/utils/node_config.go.erb [hunk 3]
```diff
@@ -102,6 +102,30 @@ func schemaNodeConfig() *schema.Schema {
 								ForceNew:         true,
 								Description: `Size of partitions to create on the GPU. Valid values are described in the NVIDIA mig user guide (https://docs.nvidia.com/datacenter/tesla/mig-user-guide/#partitioning)`,
 							},
+							"gpu_sharing_config": &schema.Schema{
+								Type:         schema.TypeList,
+								MaxItems:     1,
+								Optional:     true,
+								ForceNew:     true,
+								ConfigMode: schema.SchemaConfigModeAttr,
+								Description:  `The configuration for GPU sharing options.`,
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/utils/node_config.go.erb:111]:
```suggestion
								Description:  `Configuration for GPU sharing.`,
```

File: mmv1/third_party/terraform/utils/node_config.go.erb [hunk 4]
```diff
@@ -102,6 +102,30 @@ func schemaNodeConfig() *schema.Schema {
...
+								Elem: &schema.Resource{
+									Schema: map[string]*schema.Schema{
+										"gpu_sharing_strategy": &schema.Schema{
+											Type:        schema.TypeString,
+											Required:    true,
+											ForceNew:    true,
+											Description: `The type of GPU sharing strategy to enable on the GPU node. Possible value is TIME_SHARING`,
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/utils/node_config.go.erb:118]: Can we link to possible values instead? This documentation is likely to get outdated. markdown syntax is supported.

File: mmv1/third_party/terraform/utils/node_config.go.erb [hunk 5]
```diff
@@ -102,6 +102,30 @@ func schemaNodeConfig() *schema.Schema {
...
+										},
+										"max_shared_clients_per_gpu": &schema.Schema{
+											Type:        schema.TypeInt,
+											Required:    true,
+											ForceNew:    true,
+											Description: `The max number of containers that can share a GPU.`,
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/utils/node_config.go.erb:124]:
```suggestion
											Description: `The maximum number of containers that can share a GPU.`,
```

**Review by ScottSuarez — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/utils/node_config.go.erb (hunk 5 above)
Comment (grac3gao) [reply mmv1/third_party/terraform/utils/node_config.go.erb:124]: done

File: mmv1/third_party/terraform/utils/node_config.go.erb (hunk 3 above)
Comment (grac3gao) [reply mmv1/third_party/terraform/utils/node_config.go.erb:111]: done

File: mmv1/third_party/terraform/utils/node_config.go.erb (hunk 2 above)
Comment (grac3gao) [reply mmv1/third_party/terraform/utils/node_config.go.erb:525]: done

File: mmv1/third_party/terraform/tests/resource_container_node_pool_test.go.erb (hunk 1 above)
Comment (grac3gao) [reply mmv1/third_party/terraform/tests/resource_container_node_pool_test.go.erb:2280]: done

File: mmv1/third_party/terraform/utils/node_config.go.erb (hunk 4 above)
Comment (grac3gao) [reply mmv1/third_party/terraform/utils/node_config.go.erb]: done

**Review by ScottSuarez — APPROVED** [review]

**Comment by ScottSuarez** [comment]
Thanks Grace !

**Pull request merged**